File: src/NetFxCSharp/Facesso.GenericControls/ucTimeDetailsSettings.vb
Imports Facesso.Data
Imports System.ComponentModel
Imports System.Collections.ObjectModel

Public Class ucTimeDetailsSettings

    Protected myTSDetails As TimeSettingDetails
    Protected myCurrentlyDisplayedShift As Integer
    Protected myCurrentlyDisplayedWeekday As TimeSettingDetailsWeekdays
    Protected myWeekdayButtons As New KeyedControlCollection
    Protected myInitialized As Boolean
    Protected myHasntGotData As Boolean = True
    'Protected myShiftTabs As KeyedControlCollection

    Sub New()

        ' This call is required by the Windows Form Designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
        myTSDetails = New TimeSettingDetails()
        With myWeekdayButtons
            .Add(btnWD_01_Monday)
            .Add(btnWD_02_Tuesday)
            .Add(btnWD_03_Wednesday)
            .Add(btnWD_04_Thursday)
            .Add(btnWD_05_Friday)
            .Add(btnWD_06_Saturday)
            .Add(btnWD_07_Sunday)
        End With
        myInitialized = True
        myCurrentlyDisplayedShift = 1
        SetControls()
    End Sub

    Public Property CurrentlyDisplayedShift() As Integer
        Get
            Return myCurrentlyDisplayedShift
        End Get
        Set(ByVal value As Integer)
            If value = 0 Then value = 1
            SaveData()
            myCurrentlyDisplayedShift = value
            SetControls()
        End Set
    End Property

    Public Property CurrentlyDisplayedWeekday() As TimeSettingDetailsWeekdays
        Get
            Return myCurrentlyDisplayedWeekday
        End Get
        Set(ByVal value As TimeSettingDetailsWeekdays)
            SaveData()
            myCurrentlyDisplayedWeekday = value
            SetControls()
        End Set
    End Property

    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public Property TSDetails() As TimeSettingDetails
        Get
            Return myTSDetails
        End Get
        Set(ByVal value As TimeSettingDetails)
            If value Is Nothing Then
                myHasntGotData = True
            Else
                myHasntGotData = False
            End If
            myTSDetails = value
            SetControls()
        End Set
    End Property

    Sub SetControls()
        'Controls-Zustände anpassen
        If Not myInitialized Or myHasntGotData Then Return

        'Daten einspielen
        Dim locTSD As TimeSettingDetail
        If CurrentlyDisplayedWeekday = TimeSettingDetailsWeekdays.ForAll Then
            locTSD = TSDetails.GenericTimeSettingDetail(CurrentlyDisplayedShift - 1)
            btnReset.Enabled = True
        Else
            locTSD = TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + CurrentlyDisplayedWeekday - 1)
            btnReset.Enabled = False
        End If

        ndbCoreTimeStart.TypeSafeValue = locTSD.ShiftStart
        ndbCoreTimeEnd.TypeSafeValue = locTSD.ShiftEnd
        ndbImportTimeStart.TypeSafeValue = locTSD.ImportShiftStart
        ndbImportTimeEnd.TypeSafeValue = locTSD.ImportShiftEnd
        ndbRoundUpBefore.TypeSafeValue = locTSD.RoundUpBefore
        ndbRoundDownAfter.TypeSafeValue = locTSD.RoundDownAfter
        nibPausetime.TypeSafeValue = locTSD.WorkBreak
        nibThreshold.TypeSafeValue = locTSD.Threshold
        ncbForceToHavePause.TypeSafeValue = locTSD.ForceToHavePause

        'Dokumentation der Daten setzen
        If locTSD.ShiftStart.IsNull Or locTSD.ShiftEnd.IsNull Then
            btnStartDate.Text = "- - -"
            btnEndDate.Text = "- - -"
            lblEndTimeDateDecription.Text = ""
            lblImportEndTimeDateDescription.Text = ""
        Else
            With locTSD
                If .ShiftStart > .ShiftEnd Then
                    .ShiftEnd = .ShiftEnd.TypedValue.AddDays(1)
                End If

                If .ShiftEnd.TypedValue.Date = #1/1/2003# Then
                    lblEndTimeDateDecription.Text = "(Derselbe Tag)"
                Else
                    lblEndTimeDateDecription.Text = "(Der Folgetag)"
                End If

                btnStartDate.Text = .ShiftStart.TypedValue.ToShortDateString
                btnEndDate.Text = .ShiftEnd.TypedValue.ToShortDateString

                If Not .ImportShiftEnd.IsNull Then
                    If .ImportShiftStart > .ImportShiftEnd Then
                        .ImportShiftEnd = .ImportShiftEnd.TypedValue.AddDays(1)
                    End If


                    If .ImportShiftEnd.TypedValue.Date = #1/1/2003# Then
                        lblImportEndTimeDateDescription.Text = "(Derselbe Tag)"
                    Else
                        lblImportEndTimeDateDescription.Text = "(Der Folgetag)"
                    End If
                Else
                    .ImportShiftStart = .ShiftStart
                    .ImportShiftEnd = .ShiftEnd
                End If
            End With
        End If

        'Schichttab wählen
        Select Case CurrentlyDisplayedShift
            Case 1
                tcShifts.SelectedTab = tpShift1
                lblShiftInformer.Text = "für Schicht 1"
            Case 2
                tcShifts.SelectedTab = tpShift2
                lblShiftInformer.Text = "für Schicht 2"
            Case 3
                tcShifts.SelectedTab = tpShift3
                lblShiftInformer.Text = "für Schicht 3"
            Case 4
                tcShifts.SelectedTab = tpShift4
                lblShiftInformer.Text = "für Sonderschicht"
        End Select

        'Buttonfarbe und Selektierung setzen
        If CurrentlyDisplayedWeekday = TimeSettingDetailsWeekdays.ForAll Then
            btnGeneric.BackColor = Color.Yellow
        Else
            btnGeneric.BackColor = SystemColors.Control
        End If

        For locCount As Integer = 0 To 6
            Dim locBtn As Button = DirectCast(myWeekdayButtons(locCount), Button)
            ' Generischer Tag immer fett
            If TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount).IsDerived Then
                locBtn.Font = New Font(locBtn.Font, FontStyle.Regular)
            Else
                locBtn.Font = New Font(locBtn.Font, FontStyle.Bold)
            End If

            If locCount = Convert.ToInt32(CurrentlyDisplayedWeekday) - 1 Then
                locBtn.BackColor = Color.Yellow
            Else
                locBtn.BackColor = SystemColors.Control
            End If
        Next

        'Listbox aufbereiten:
        'Erst die Generischen Tage anzeigen
        lbTimes.Items.Clear()
        For z As Integer = 0 To 3
            lbTimes.Items.Add(Me.TSDetails.GenericTimeSettingDetail(z).ToString)
        Next

        'Dann den Rest
        For z As Integer = 0 To 6
            For s As Integer = 0 To 3
                lbTimes.Items.Add(Me.TSDetails.TimeSettingDetail((s) * 7 + z))
            Next
        Next


    End Sub

    Private Sub SaveData()
        Dim locTSD As TimeSettingDetail
        Dim locIsGeneric As Boolean = False
        If CurrentlyDisplayedWeekday = TimeSettingDetailsWeekdays.ForAll Then
            locTSD = TSDetails.GenericTimeSettingDetail(CurrentlyDisplayedShift - 1)
            locIsGeneric = True
        Else
            locTSD = TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + CurrentlyDisplayedWeekday - 1)
        End If

        locTSD.ShiftStart = ndbCoreTimeStart.TypeSafeValue
        locTSD.ShiftEnd = ndbCoreTimeEnd.TypeSafeValue
        locTSD.ImportShiftStart = ndbImportTimeStart.TypeSafeValue
        locTSD.ImportShiftEnd = ndbImportTimeEnd.TypeSafeValue
        locTSD.RoundUpBefore = ndbRoundUpBefore.TypeSafeValue
        locTSD.RoundDownAfter = ndbRoundDownAfter.TypeSafeValue
        locTSD.WorkBreak = nibPausetime.TypeSafeValue
        locTSD.Threshold = nibThreshold.TypeSafeValue
        locTSD.ForceToHavePause = ncbForceToHavePause.TypeSafeValue
        locTSD.ForShift = CurrentlyDisplayedShift

        'Plausibilitätskontrolle
        If locTSD.ShiftStart.IsNull Or locTSD.ShiftEnd.IsNull Then
            locTSD.NullAll()
        Else
            With locTSD
                Dim locTimeSpanStart As TimeSpan
                Dim locTimeSpanEnd As TimeSpan

                If Not .ImportShiftStart.IsNull Then
                    locTimeSpanStart = .ImportShiftStart.TypedValue.TimeOfDay
                    .ImportShiftStart = #1/1/2003#.Add(locTimeSpanStart)
                End If

                If Not .ImportShiftEnd.IsNull Then
                    locTimeSpanEnd = .ImportShiftEnd.TypedValue.TimeOfDay
                    .ImportShiftEnd = #1/1/2003#.Add(locTimeSpanEnd)
                End If

                locTimeSpanStart = .ShiftStart.TypedValue.TimeOfDay
                locTimeSpanEnd = .ShiftEnd.TypedValue.TimeOfDay
                .ShiftStart = #1/1/2003#.Add(locTimeSpanStart)
                .ShiftEnd = #1/1/2003#.Add(locTimeSpanEnd)

                If .ShiftStart > .ShiftEnd Then
                    .ShiftEnd = .ShiftEnd.TypedValue.AddDays(1)
                End If

                If .ImportShiftStart.IsNull Then
                    .ImportShiftStart = .ShiftStart.TypedValue
                End If

                If .ImportShiftEnd.IsNull Then
                    .ImportShiftEnd = .ShiftEnd.TypedValue
                End If

                If .ImportShiftStart > .ImportShiftEnd Then
                    .ImportShiftEnd = .ImportShiftEnd.TypedValue.AddDays(1)
                End If
            End With
        End If

        With locTSD
            If Not .RoundUpBefore.IsNull Then
                Dim locTimeSpan As TimeSpan = .RoundUpBefore.TypedValue.TimeOfDay
                .RoundUpBefore = .ImportShiftStart.TypedValue.Date
                .RoundUpBefore = .RoundUpBefore.TypedValue.Add(locTimeSpan)
                If .RoundUpBefore > .ImportShiftStart Then
                    .RoundUpBefore = .ImportShiftStart
                End If
            End If

            If Not .RoundDownAfter.IsNull Then
                Dim locTimeSpan As TimeSpan = .RoundDownAfter.TypedValue.TimeOfDay
                .RoundDownAfter = .ImportShiftEnd.TypedValue.Date
                .RoundDownAfter = .RoundDownAfter.TypedValue.Add(locTimeSpan)
                If .RoundDownAfter < .ImportShiftEnd Then
                    .RoundDownAfter = .ImportShiftEnd
                End If
            End If

        End With

        'Falls es kein generischer Datenblock war, dann rausfinden, ob
        'Daten der generischen Vorlage entsprechen!
        If Not locIsGeneric Then
            If TSDetails.GenericTimeSettingDetail(CurrentlyDisplayedShift - 1).IsEqual(locTSD) Then
                locTSD.IsDerived = True
            Else
                locTSD.IsDerived = False
            End If
        End If

        'Falls Änderungen an der Vorlage, dann alle ändern,
        'die nicht abgeleitet sind!
        If locIsGeneric Then
            For locCount As Integer = 0 To 6
                If TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount).IsDerived Then
                    TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount) = locTSD.Clone
                    TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount).ForWeekday = CType(locCount + 1, TimeSettingDetailsWeekdays)
                End If
                'Und die, die gleich geworden sind, wieder einbinden.
                If TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount).IsEqual(locTSD) Then
                    TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount).IsDerived = True
                End If
            Next
        End If
    End Sub

    Private Sub btnGeneric_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGeneric.Click, btnWD_01_Monday.Click, btnWD_07_Sunday.Click, btnWD_06_Saturday.Click, btnWD_05_Friday.Click, btnWD_04_Thursday.Click, btnWD_03_Wednesday.Click, btnWD_02_Tuesday.Click
        If DirectCast(sender, Control).Name = btnGeneric.Name Then
            CurrentlyDisplayedWeekday = TimeSettingDetailsWeekdays.ForAll
        Else
            For i As Integer = 0 To 6
                If DirectCast(sender, Control).Name = myWeekdayButtons(i).Name Then
                    CurrentlyDisplayedWeekday = CType(i + 1, TimeSettingDetailsWeekdays)
                    Return
                End If
            Next
        End If
    End Sub

    Private Sub tcShifts_Selected(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TabControlEventArgs) Handles tcShifts.Selected
        CurrentlyDisplayedShift = e.TabPageIndex + 1
    End Sub

    Private Sub ndbCoreTimeStart_Validated(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ndbCoreTimeStart.Validated
        If ndbCoreTimeEnd.TypeSafeValue.IsNull Then
            ndbCoreTimeEnd.TypeSafeValue = ndbCoreTimeStart.TypeSafeValue
        End If
        SaveData()
        SetControls()
    End Sub

    Private Sub ndbCoreTimeEnd_Validated(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ndbCoreTimeEnd.Validated
        If ndbCoreTimeStart.TypeSafeValue.IsNull Then
            ndbCoreTimeStart.TypeSafeValue = ndbCoreTimeEnd.TypeSafeValue
        End If
        SaveData()
        SetControls()
    End Sub

    Private Sub ndbRoundUpBefore_Validated(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ndbRoundUpBefore.Validated, ndbRoundDownAfter.Validated
        SaveData()
        SetControls()
    End Sub

    Private Sub ndbImportTimeStart_Validated(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ndbImportTimeStart.Validated
        If ndbImportTimeEnd.TypeSafeValue.IsNull Then
            ndbImportTimeEnd.TypeSafeValue = ndbImportTimeStart.TypeSafeValue
        End If
        SaveData()
        SetControls()
    End Sub

    Private Sub ndbImportTimeEnd_Validated(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ndbImportTimeEnd.Validated
        If ndbImportTimeStart.TypeSafeValue.IsNull Then
            ndbImportTimeStart.TypeSafeValue = ndbImportTimeEnd.TypeSafeValue
        End If
        SaveData()
        SetControls()
    End Sub

    ''' <summary>
    ''' Setzt, wenn "für alle Wochentage" ausgewählt ist, alle Tagesunterschiede zurück.
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub btnReset_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnReset.Click
        For locCount As Integer = 0 To 6
            TSDetails.TimeSettingDetail((CurrentlyDisplayedShift - 1) * 7 + locCount) = TSDetails.GenericTimeSettingDetail(CurrentlyDisplayedShift - 1).Clone
            SetControls()
        Next
    End Sub
End Class

Public Class KeyedControlCollection
    Inherits KeyedCollection(Of String, Control)

    Protected Overrides Function GetKeyForItem(ByVal item As System.Windows.Forms.Control) As String
        Return item.Name
    End Function
End Class

File: src/NetFxCSharp/Facesso.GenericControls/ucTimeDetailsSettings.cs
using Facesso.Data;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.GenericControls
{
    public partial class ucTimeDetailsSettings
    {
        protected TimeSettingDetails myTSDetails;
        protected int myCurrentlyDisplayedShift;
        protected TimeSettingDetailsWeekdays myCurrentlyDisplayedWeekday;
        protected KeyedControlCollection myWeekdayButtons = new KeyedControlCollection();
        protected bool myInitialized;
        protected bool myHasntGotData = true;
        //Protected myShiftTabs As KeyedControlCollection
        public ucTimeDetailsSettings()
        {
            // This call is required by the Windows Form Designer.
            InitializeComponent();
            // Add any initialization after the InitializeComponent() call.
            myTSDetails = new TimeSettingDetails();
            {
                var __with0 = myWeekdayButtons;
                __with0.Add(btnWD_01_Monday);
                __with0.Add(btnWD_02_Tuesday);
                __with0.Add(btnWD_03_Wednesday);
                __with0.Add(btnWD_04_Thursday);
                __with0.Add(btnWD_05_Friday);
                __with0.Add(btnWD_06_Saturday);
                __with0.Add(btnWD_07_Sunday);
            }

            myInitialized = true;
            myCurrentlyDisplayedShift = 1;
            SetControls();
        }

        public int CurrentlyDisplayedShift
        {
            get
            {
                return myCurrentlyDisplayedShift;
            }

            set
            {
                if (value == 0)
                {
                    value = 1;
                }

                SaveData();
                myCurrentlyDisplayedShift = value;
                SetControls();
            }
        }

        public TimeSettingDetailsWeekdays CurrentlyDisplayedWeekday
        {
            get
            {
                return myCurrentlyDisplayedWeekday;
            }

            set
            {
                SaveData();
                myCurrentlyDisplayedWeekday = value;
                SetControls();
            }
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public TimeSettingDetails TSDetails
        {
            get
            {
                return myTSDetails;
            }

            set
            {
                if (value == null)
                {
                    myHasntGotData = true;
                }
                else
                {
                    myHasntGotData = false;
                }

                myTSDetails = value;
                SetControls();
            }
        }

        public void SetControls()
        {
            //Controls-Zustände anpassen
            if (!(myInitialized) | myHasntGotData)
            {
                return;
            }

            //Daten einspielen
            TimeSettingDetail locTSD = default(TimeSettingDetail);
            if (CurrentlyDisplayedWeekday == TimeSettingDetailsWeekdays.ForAll)
            {
                locTSD = TSDetails.GenericTimeSettingDetail[CurrentlyDisplayedShift - 1];
                btnReset.Enabled = true;
            }
            else
            {
                locTSD = TSDetails.TimeSettingDetail[(int)((CurrentlyDisplayedShift - 1) * 7 + CurrentlyDisplayedWeekday - 1)];
                btnReset.Enabled = false;
            }

            ndbCoreTimeStart.TypeSafeValue = locTSD.ShiftStart;
            ndbCoreTimeEnd.TypeSafeValue = locTSD.ShiftEnd;
            ndbImportTimeStart.TypeSafeValue = locTSD.ImportShiftStart;
            ndbImportTimeEnd.TypeSafeValue = locTSD.ImportShiftEnd;
            ndbRoundUpBefore.TypeSafeValue = locTSD.RoundUpBefore;
            ndbRoundDownAfter.TypeSafeValue = locTSD.RoundDownAfter;
            nibPausetime.TypeSafeValue = locTSD.WorkBreak;
            nibThreshold.TypeSafeValue = locTSD.Threshold;
            ncbForceToHavePause.TypeSafeValue = locTSD.ForceToHavePause;
            //Dokumentation der Daten setzen
            if (locTSD.ShiftStart.IsNull | locTSD.ShiftEnd.IsNull)
            {
                btnStartDate.Text = "- - -";
                btnEndDate.Text = "- - -";
                lblEndTimeDateDecription.Text = "";
                lblImportEndTimeDateDescription.Text = "";
            }
            else
            {
                {
                    var __with1 = locTSD;
                    if (__with1.ShiftStart > __with1.ShiftEnd)
                    {
                        __with1.ShiftEnd = __with1.ShiftEnd.TypedValue.AddDays(1);
                    }

                    if (__with1.ShiftEnd.TypedValue.Date == new System.DateTime(2003, 1, 1))
                    {
                        lblEndTimeDateDecription.Text = "(Derselbe Tag)";
                    }
                    else
                    {
                        lblEndTimeDateDecription.Text = "(Der Folgetag)";
                    }

                    btnStartDate.Text = __with1.ShiftStart.TypedValue.ToShortDateString();
                    btnEndDate.Text = __with1.ShiftEnd.TypedValue.ToShortDateString();
                    if (!(__with1.ImportShiftEnd.IsNull))
                    {
                        if (__with1.ImportShiftStart > __with1.ImportShiftEnd)
                        {
                            __with1.ImportShiftEnd = __with1.ImportShiftEnd.TypedValue.AddDays(1);
                        }

                        if (__with1.ImportShiftEnd.TypedValue.Date == new System.DateTime(2003, 1, 1))
                        {
                            lblImportEndTimeDateDescription.Text = "(Derselbe Tag)";
                        }
                        else
                        {
                            lblImportEndTimeDateDescription.Text = "(Der Folgetag)";
                        }
                    }
                    else
                    {
                        __with1.ImportShiftStart = __with1.ShiftStart;
                        __with1.ImportShiftEnd = __with1.ShiftEnd;
                    }
                }
            }

            //Schichttab wählen
            {
                var __select2 = CurrentlyDisplayedShift;
                if (__select2 == 1)
                {
                    tcShifts.SelectedTab = tpShift1;
                    lblShiftInformer.Text = "für Schicht 1";
                }
                else if (__select2 == 2)
                {
                    tcShifts.SelectedTab = tpShift2;
                    lblShiftInformer.Text = "für Schicht 2";
                }
                else if (__select2 == 3)
                {
                    tcShifts.SelectedTab = tpShift3;
                    lblShiftInformer.Text = "für Schicht 3";
                }
                else if (__select2 == 4)
                {
                    tcShifts.SelectedTab = tpShift4;
                    lblShiftInformer.Text = "für Sonderschicht";
                }
            }

            //Buttonfarbe und Selektierung setzen
            if (CurrentlyDisplayedWeekday == TimeSettingDetailsWeekdays.ForAll)
            {
                btnGeneric.BackColor = Color.Yellow;
            }
            else
            {
                btnGeneric.BackColor = SystemColors.Control;
            }

            for (int locCount = 0; locCount <= 6; locCount++)
            {
                Button locBtn = ((Button)myWeekdayButtons[locCount]);
                // Generischer Tag immer fett
                if (TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount].IsDerived)
                {
                    locBtn.Font = new Font(locBtn.Font, FontStyle.Regular);
                }
                else
                {
                    locBtn.Font = new Font(locBtn.Font, FontStyle.Bold);
                }

                if (locCount == Convert.ToInt32(CurrentlyDisplayedWeekday) - 1)
                {
                    locBtn.BackColor = Color.Yellow;
                }
                else
                {
                    locBtn.BackColor = SystemColors.Control;
                }
            }

            //Listbox aufbereiten:
            //Erst die Generischen Tage anzeigen
            lbTimes.Items.Clear();
            for (int z = 0; z <= 3; z++)
            {
                lbTimes.Items.Add(this.TSDetails.GenericTimeSettingDetail[z].ToString());
            }

            //Dann den Rest
            for (int z = 0; z <= 6; z++)
            {
                for (int s = 0; s <= 3; s++)
                {
                    lbTimes.Items.Add(this.TSDetails.TimeSettingDetail[(s) * 7 + z]);
                }
            }
        }

        private void SaveData()
        {
            TimeSettingDetail locTSD = default(TimeSettingDetail);
            bool locIsGeneric = false;
            if (CurrentlyDisplayedWeekday == TimeSettingDetailsWeekdays.ForAll)
            {
                locTSD = TSDetails.GenericTimeSettingDetail[CurrentlyDisplayedShift - 1];
                locIsGeneric = true;
            }
            else
            {
                locTSD = TSDetails.TimeSettingDetail[(int)((CurrentlyDisplayedShift - 1) * 7 + CurrentlyDisplayedWeekday - 1)];
            }

            locTSD.ShiftStart = ndbCoreTimeStart.TypeSafeValue;
            locTSD.ShiftEnd = ndbCoreTimeEnd.TypeSafeValue;
            locTSD.ImportShiftStart = ndbImportTimeStart.TypeSafeValue;
            locTSD.ImportShiftEnd = ndbImportTimeEnd.TypeSafeValue;
            locTSD.RoundUpBefore = ndbRoundUpBefore.TypeSafeValue;
            locTSD.RoundDownAfter = ndbRoundDownAfter.TypeSafeValue;
            locTSD.WorkBreak = nibPausetime.TypeSafeValue;
            locTSD.Threshold = nibThreshold.TypeSafeValue;
            locTSD.ForceToHavePause = ncbForceToHavePause.TypeSafeValue;
            locTSD.ForShift = CurrentlyDisplayedShift;
            //Plausibilitätskontrolle
            if (locTSD.ShiftStart.IsNull | locTSD.ShiftEnd.IsNull)
            {
                locTSD.NullAll();
            }
            else
            {
                {
                    var __with3 = locTSD;
                    TimeSpan locTimeSpanStart = default(TimeSpan);
                    TimeSpan locTimeSpanEnd = default(TimeSpan);
                    if (!(__with3.ImportShiftStart.IsNull))
                    {
                        locTimeSpanStart = __with3.ImportShiftStart.TypedValue.TimeOfDay;
                        __with3.ImportShiftStart = new System.DateTime(2003, 1, 1).Add(locTimeSpanStart);
                    }

                    if (!(__with3.ImportShiftEnd.IsNull))
                    {
                        locTimeSpanEnd = __with3.ImportShiftEnd.TypedValue.TimeOfDay;
                        __with3.ImportShiftEnd = new System.DateTime(2003, 1, 1).Add(locTimeSpanEnd);
                    }

                    locTimeSpanStart = __with3.ShiftStart.TypedValue.TimeOfDay;
                    locTimeSpanEnd = __with3.ShiftEnd.TypedValue.TimeOfDay;
                    __with3.ShiftStart = new System.DateTime(2003, 1, 1).Add(locTimeSpanStart);
                    __with3.ShiftEnd = new System.DateTime(2003, 1, 1).Add(locTimeSpanEnd);
                    if (__with3.ShiftStart > __with3.ShiftEnd)
                    {
                        __with3.ShiftEnd = __with3.ShiftEnd.TypedValue.AddDays(1);
                    }

                    if (__with3.ImportShiftStart.IsNull)
                    {
                        __with3.ImportShiftStart = __with3.ShiftStart.TypedValue;
                    }

                    if (__with3.ImportShiftEnd.IsNull)
                    {
                        __with3.ImportShiftEnd = __with3.ShiftEnd.TypedValue;
                    }

                    if (__with3.ImportShiftStart > __with3.ImportShiftEnd)
                    {
                        __with3.ImportShiftEnd = __with3.ImportShiftEnd.TypedValue.AddDays(1);
                    }
                }
            }

            {
                var __with4 = locTSD;
                if (!(__with4.RoundUpBefore.IsNull))
                {
                    TimeSpan locTimeSpan = __with4.RoundUpBefore.TypedValue.TimeOfDay;
                    __with4.RoundUpBefore = __with4.ImportShiftStart.TypedValue.Date;
                    __with4.RoundUpBefore = __with4.RoundUpBefore.TypedValue.Add(locTimeSpan);
                    if (__with4.RoundUpBefore > __with4.ImportShiftStart)
                    {
                        __with4.RoundUpBefore = __with4.ImportShiftStart;
                    }
                }

                if (!(__with4.RoundDownAfter.IsNull))
                {
                    TimeSpan locTimeSpan = __with4.RoundDownAfter.TypedValue.TimeOfDay;
                    __with4.RoundDownAfter = __with4.ImportShiftEnd.TypedValue.Date;
                    __with4.RoundDownAfter = __with4.RoundDownAfter.TypedValue.Add(locTimeSpan);
                    if (__with4.RoundDownAfter < __with4.ImportShiftEnd)
                    {
                        __with4.RoundDownAfter = __with4.ImportShiftEnd;
                    }
                }
            }

            //Falls es kein generischer Datenblock war, dann rausfinden, ob
            //Daten der generischen Vorlage entsprechen!
            if (!(locIsGeneric))
            {
                if (TSDetails.GenericTimeSettingDetail[CurrentlyDisplayedShift - 1].IsEqual(locTSD))
                {
                    locTSD.IsDerived = true;
                }
                else
                {
                    locTSD.IsDerived = false;
                }
            }

            //Falls Änderungen an der Vorlage, dann alle ändern,
            //die nicht abgeleitet sind!
            if (locIsGeneric)
            {
                for (int locCount = 0; locCount <= 6; locCount++)
                {
                    if (TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount].IsDerived)
                    {
                        TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount] = locTSD.Clone();
                        TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount].ForWeekday = ((TimeSettingDetailsWeekdays)locCount + 1);
                    }

                    //Und die, die gleich geworden sind, wieder einbinden.
                    if (TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount].IsEqual(locTSD))
                    {
                        TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount].IsDerived = true;
                    }
                }
            }
        }

        private void btnGeneric_Click(System.Object sender, System.EventArgs e)
        {
            if (((Control)sender).Name == btnGeneric.Name)
            {
                CurrentlyDisplayedWeekday = TimeSettingDetailsWeekdays.ForAll;
            }
            else
            {
                for (int i = 0; i <= 6; i++)
                {
                    if (((Control)sender).Name == myWeekdayButtons[i].Name)
                    {
                        CurrentlyDisplayedWeekday = ((TimeSettingDetailsWeekdays)i + 1);
                        return;
                    }
                }
            }
        }

        private void tcShifts_Selected(System.Object sender, System.Windows.Forms.TabControlEventArgs e)
        {
            CurrentlyDisplayedShift = e.TabPageIndex + 1;
        }

        private void ndbCoreTimeStart_Validated(System.Object sender, System.EventArgs e)
        {
            if (ndbCoreTimeEnd.TypeSafeValue.IsNull)
            {
                ndbCoreTimeEnd.TypeSafeValue = ndbCoreTimeStart.TypeSafeValue;
            }

            SaveData();
            SetControls();
        }

        private void ndbCoreTimeEnd_Validated(System.Object sender, System.EventArgs e)
        {
            if (ndbCoreTimeStart.TypeSafeValue.IsNull)
            {
                ndbCoreTimeStart.TypeSafeValue = ndbCoreTimeEnd.TypeSafeValue;
            }

            SaveData();
            SetControls();
        }

        private void ndbRoundUpBefore_Validated(System.Object sender, System.EventArgs e)
        {
            SaveData();
            SetControls();
        }

        private void ndbImportTimeStart_Validated(System.Object sender, System.EventArgs e)
        {
            if (ndbImportTimeEnd.TypeSafeValue.IsNull)
            {
                ndbImportTimeEnd.TypeSafeValue = ndbImportTimeStart.TypeSafeValue;
            }

            SaveData();
            SetControls();
        }

        private void ndbImportTimeEnd_Validated(System.Object sender, System.EventArgs e)
        {
            if (ndbImportTimeStart.TypeSafeValue.IsNull)
            {
                ndbImportTimeStart.TypeSafeValue = ndbImportTimeEnd.TypeSafeValue;
            }

            SaveData();
            SetControls();
        }

        /// <summary>
        /// Setzt, wenn "für alle Wochentage" ausgewählt ist, alle Tagesunterschiede zurück.
        /// </summary>
        /// <param name = "sender"></param>
        /// <param name = "e"></param>
        /// <remarks></remarks>
        private void btnReset_Click(System.Object sender, System.EventArgs e)
        {
            for (int locCount = 0; locCount <= 6; locCount++)
            {
                TSDetails.TimeSettingDetail[(CurrentlyDisplayedShift - 1) * 7 + locCount] = TSDetails.GenericTimeSettingDetail[CurrentlyDisplayedShift - 1].Clone();
                SetControls();
            }
        }
    }

    public class KeyedControlCollection : KeyedCollection<string, Control>
    {
        protected override string GetKeyForItem(System.Windows.Forms.Control item)
        {
            return item.Name;
        }
    }
}